### DISCUSSION TITLE
Taking account of selected action in next action space

### DISCUSSION THREAD

**Discussion by DonalOCois** [JuliaPOMDP/POMDPs.jl#455] [Problem Implementation] [answered]
Hi all,
I have been working on something for a few days now that I hope is somehow doable, but I haven't found a solution.
Basically what I want to do is to take account of the action that I have selected in which actions will be available in the next time step.

So, my code is as follows. I have 4 elements in my state space, the first two representing the position of a robot on an x-y axis and the latter two representing an internal state , which is based on whatever action was selected. I think of it like steering your car to the left being the action, and the internal state variables here (isx and isy) representing the position of your steering wheel.
In the actions function, the available action space gets updated on the basis of those internal and also external state variables.

```
mdp = QuickMDP(
	function gen(s, a, rng)
		x, y, isx, isy = s
		xₚ = clamp(x+a[1], -10, 10)
		yₚ = clamp(y+a[2], -10, 10)
                isxₚ = clamp(isx+a[1], -1, 1)
		isyₚ = clamp(isy+a[2], -1, 1)
return (sp=[xₚ, yₚ, isxₚ, isyₚ])	
	end,



actions = function ((a=[0,0]),(s=(0.,0.,1.,0.)))
if s[3] == 1
         if s[1] >= 5
                  return [[-1,1], [-1,-1]]
         else 
                 return [[1,0]]
end
elseif.. _(other possible movements)_
end
end,
```

This doesn't seem to work unfortunately. To test it, I choose s[3] to be 1 by default as you see. Therefore the first if statement is true. The second if statement should also become true when the position of the robot on the x-axis (s[1]) reaches 5. This does not happen, and instead the code carries on returning [1,0] as the action space. In other words, after the first iteration, the action space does not update anymore (is not returned by the second snippet), to take account of the changed state space.

So my question is simply if anyone sees something wrong with my approach here. If anyone does have a suggestion (which would be hugely appreciated of course), please do give the most explicit detail you can. Sadly, if there is a mistake to be made, I am quite capable of finding it!
Thanks very much
Donal

**Comment by zsunberg** [comment, marked as answer]
Hi @DonalOCois , I think you have the right idea - you should augment the state space to include the previous action if the current action space depends on the previous action (the steering wheel you mention is the right idea).

However, it doesn't look like you are implementing it quite right. It might be easier to debug and the syntax might be easier to understand if you break the function out separately from the `QuickMDP` constructor like this:

```
function my_actions(s = #=default state=#)
    # return the action space
end

mdp = QuickMDP(
    #...
    actions = my_actions,
    #...
)
```
Then you can debug the `my_actions` function more easily. Note that the function should only take one argument that is the state.

**Reply by DonalOCois** [reply]
Hey @zsunberg , thanks again for helping me out.
I had thought the same thing about the function taking only one argument - the state - but when I do that, I get the following problem. Maybe you've seen it before and know how to deal with it.

If this is the "my_actions" function as you suggested, with just one simple if statement that should change the action space after 5 time steps.
```
function my_actions(s = (0.,0.,1.,0.))
if s[1] > 5.0
					return [[-1,1], [-1,-1]]
		else
					return [[1,0]]
		end
end
```
We get an error as seen below when we try to send it to solve for a policy. It wants the "key" of all possible actions, and not only the action space that we pass to it.

![image](https://user-images.githubusercontent.com/106449696/215065578-081bbb01-40ae-4bad-9d4e-2bcccec9d48e.png)

If instead, we pass a default action to my_actions like this for example `function my_actions(a=(nothing),s = (0.,0.,1.,0.))`, we no longer have this "key" error, but we have the problem I described previously.

My intuition is that what's happening in that latter case is that this solver takes a snap-shot of the action space at the start, and then uses the 'solve' function to form a policy, given that snapshot. So it will never take subsequent changes to the action space into account. That is at-odds with my understanding of an MDP policy as being the mapping of every state to an action, which must rely on the available action space in every state, so I guess I'm just making some pretty fundamental error.

**Reply by zsunberg** [reply]
You need a branch in your function that returns *all* of the actions of no state argument is given

**Reply by DonalOCois** [reply]
Thanks, you were right! That did it. Solver handles it fine now.

**Reply by DonalOCois** [reply]
Sorry to bother you again so soon @zsunberg, but I have a directly related follow-up question.
The local approximation value iteration solver works great if I give it the entire action space from the beginning. It immediately takes the required action to find the highest reward state. However, in my implementation, because I limit the action space at the early stages of the simulation (it can only choose from a subset of actions), and then give it full access to the action space later, it just gets lost completely at that point. It doesn't even attempt to reach higher value states.

My intuition is that this is because it has been forced, by lack of available options, to take many sub-optimal actions in the time steps prior to having full access to the full action space. Because it uses the Bellman equation over the entire number of time steps in the simulation, not being able to choose as it wishes for a period of the simulation makes it unable to calculate properly. If there was a way to make it only calculate over a given subset of time steps, this would be ideal.
 
I think I might have more luck trying other solvers like sarsa for this reason, which can calculate over a given adjustable number of steps, but I note that they don't work with the 'generated' state and action spaces right away. I gather that they require the use of "@ gen" to discretise the states, rather than using interpolation, but I can't find implementations of this with syntax that is clear to me. I was hoping you could help me there, if you think I've analysed the problem correctly. 

If I basically had a standard QuickMDP like:
```
mdp = QuickMDP(
function gen(s, a, rng)
....
)
```

I would need to bring it to the point of using it in a solver like:
```
solver = SARSASolver(exppolicy, learning_rate=0.1, n_episodes=5000, max_episode_length=50, eval_every=50, n_eval_traj=100)
policy = solve(solver, mdp)
```

**Reply by zsunberg** [reply]
Hi @DonalOCois , It is a bit hard to understand exactly what you're saying, but usually it is best to stick strictly to the MDP formalism and compose smaller pieces that already exist to create something more complex. Would the following approach work?

1. Create MDP1 with the limited action space.
2. Solve MDP1 to get policy1
3. Create MDP2 with the full action space
4. Solve MDP2 to get policy2
5. Use policy1 in the first part of your simulation until all actions are available, then switch to policy 2.

**Reply by DonalOCois** [reply]
Hi @zsunberg, thanks for that suggestion - it solves part of my problem.
The other problem I was having, which I didn't fully understand at the time and couldn't articulate clearly, was the high dimensionality of my state space.
While I have managed to get better agent behaviour using Temporal Difference solvers like SARSA with a discrete state space relative to the value iteration solver, the state space in the former case is very sparse. 
I wanted to ask if it's possible to use a generative state space for this instead, as I assume this would result in a less sparse state space, given that states are generated as they arise, rather than the state space consisting of every possible variation.

I noted the following documentation, which seemed to indicate that "Solvers and simulators should call [@ gen] rather than the gen function"
https://juliapomdp.github.io/POMDPs.jl/latest/api/#POMDPs.@gen
![image](https://user-images.githubusercontent.com/106449696/217508683-f5c3da53-0655-484b-bd3e-f8a3fec25f21.png)
 
But I wasn't able to get this to work, so I was hoped you might be able to demonstrate the correct syntax given what I wrote at the end of my prior post. If I've misinterpreted this and it's not possible, then no worries and thanks again for all your help.

**Reply by zsunberg** [reply]
This example of mountaincar implements a generative model: https://github.com/JuliaPOMDP/QuickPOMDPs.jl/blob/master/examples/mountaincar.jl. The first argument of a `QuickMDP` is interpreted as the `gen` function if no keyword is given. Then a solver or simulator could, for example call `@gen(:sp)(mountaincar, (0.0, 0.0), -1.0)` to get the next state. Does that answer your question?

**Reply by DonalOCois** [reply]
Well I can see how that returns the next state, given the specific states and actions that you pass to it. I can also see how you might use a for loop to iterate over a bunch of possible state-action pairs and get 'sp' returned for each of them.

But, and sorry if it should be obvious, it's not clear to me how that's used with solvers, which seem to require the mdp itself to be passed to them so that they can iterate over many possible state-action pairs. I.e something like: `policy = solve(solver, mountaincar)` 
Could you show how a solver would do it in this case?

**Reply by zsunberg** [reply]
Ah, I think I might understand what you're getting at. Are you saying that it's only possible to reach a small number of states? For example, if the state is represented by a real number, but it is only possible to reach 1.1, 2.3, 4.7, and 6.1 ?

**Reply by DonalOCois** [reply]
Simpler than that - it's just a syntax issue. You said that a "solver or simulator could, for example call @gen(:sp)(mountaincar, (0.0, 0.0), -1.0) to get the next state". I'm just asking how that would explicitly be done, because in examples that I have found, the "solve" function calls a given solver and MDP.

**Reply by zsunberg** [reply]
Here is an example of where it is used in MCTS: https://github.com/JuliaPOMDP/MCTS.jl/blob/397195917faf35d02aa9ffe13eaec428fd7c1861/src/vanilla.jl#L287 

Note that since MCTS is an *online* solver, no actual work is carried out in the `solve(::MCTSSolver, ::MDP)` function. All of the planning is done online in the `action(::MCTSPlanner, state)` function. That way `@gen` is only called on states that are reached within the MCTS tree, so there is no need to enumerate all of the state space.
